File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        map<int, bool> m;
        int left = 0;
        int length = 0;
        int maxLength = 0;
        for(int i = 0; i < 26; i++){
            m[i] = false;
        }
        for(int right = 0; right < s.length(); right++){
            while(m[s[right]-'a']){
                length--;
                m[s[left]-'a'] = false;
                left++;
            }
            m[s[right]-'a'] = true;
            length++;
            maxLength = max(length, maxLength);
        }
        return maxLength;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        letters = {}
        longest = 0
        l = 0  # Initialize left pointer
        
        for r in range(len(s)):
            if s[r] in letters:
                l = max(l, letters[s[r]] + 1)  # Move left pointer to avoid duplicate
            
            letters[s[r]] = r  # Update the index of the character
            longest = max(longest, r - l + 1)  # Calculate the max length
        
        return longest
